File: src/codewhispererChat/controllers/chat/telemetryHelper.ts
/*!
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import { UserIntent } from '@amzn/codewhisperer-streaming'
import {
    AmazonqAddMessage,
    AmazonqInteractWithMessage,
    CwsprChatCommandType,
    CwsprChatTriggerInteraction,
    CwsprChatUserIntent,
    MetricBase,
    Result,
    telemetry,
} from '../../../shared/telemetry/telemetry'
import { ChatSessionStorage } from '../../storages/chatSession'
import {
    ChatItemFeedbackMessage,
    ChatItemVotedMessage,
    CopyCodeToClipboard,
    FooterInfoLinkClick,
    InsertCodeAtCursorPosition,
    PromptAnswer,
    PromptMessage,
    ResponseBodyLinkClickMessage,
    SourceLinkClickMessage,
    TriggerPayload,
} from './model'
import { TriggerEvent, TriggerEventsStorage } from '../../storages/triggerEvents'
import globals from '../../../shared/extensionGlobals'
import { getLogger } from '../../../shared/logger'
import { codeWhispererClient } from '../../../codewhisperer/client/codewhisperer'
import CodeWhispererUserClient, { Dimension } from '../../../codewhisperer/client/codewhispereruserclient'
import { isAwsError } from '../../../shared/errors'

const performance = globalThis.performance ?? require('perf_hooks').performance

export function mapToClientTelemetryEvent<T extends MetricBase>(
    name: string,
    event: T
): CodeWhispererUserClient.SendTelemetryEventRequest {
    return {
        telemetryEvent: {
            metricData: {
                metricName: name,
                metricValue: 1,
                timestamp: new Date(Date.now()),
                dimensions: Object.keys(event).map(
                    (key): Dimension => ({ name: key, value: event[key as keyof T]?.toString() })
                ),
            },
        },
    }
}

export function logSendTelemetryEventFailure(error: any) {
    let requestId: string | undefined
    if (isAwsError(error)) {
        requestId = error.requestId
    }

    getLogger().debug(
        `Failed to sendTelemetryEvent to CodeWhisperer, requestId: ${requestId ?? ''}, message: ${error.message}`
    )
}

export function recordTelemetryChatRunCommand(type: CwsprChatCommandType, command?: string) {
    telemetry.amazonq_runCommand.emit({
        result: 'Succeeded',
        cwsprChatCommandType: type,
        cwsprChatCommandName: command,
    })
}

export class CWCTelemetryHelper {
    private sessionStorage: ChatSessionStorage
    private triggerEventsStorage: TriggerEventsStorage
    private responseStreamStartTime: Map<string, number> = new Map()
    private responseStreamTotalTime: Map<string, number> = new Map()
    private responseStreamTimeForChunks: Map<string, number[]> = new Map()

    constructor(sessionStorage: ChatSessionStorage, triggerEventsStorage: TriggerEventsStorage) {
        this.sessionStorage = sessionStorage
        this.triggerEventsStorage = triggerEventsStorage
    }

    private getUserIntentForTelemetry(userIntent: UserIntent | undefined): CwsprChatUserIntent | undefined {
        switch (userIntent) {
            case UserIntent.EXPLAIN_CODE_SELECTION:
                return 'explainCodeSelection'
            case UserIntent.SUGGEST_ALTERNATE_IMPLEMENTATION:
                return 'suggestAlternateImplementation'
            case UserIntent.APPLY_COMMON_BEST_PRACTICES:
                return 'applyCommonBestPractices'
            case UserIntent.IMPROVE_CODE:
                return 'improveCode'
            case UserIntent.CITE_SOURCES:
                return 'citeSources'
            case UserIntent.EXPLAIN_LINE_BY_LINE:
                return 'explainLineByLine'
            case UserIntent.SHOW_EXAMPLES:
                return 'showExample'
            default:
                return undefined
        }
    }

    public recordOpenChat() {
        telemetry.amazonq_openChat.emit({ result: 'Succeeded' })
    }

    public recordCloseChat() {
        telemetry.amazonq_closeChat.emit({ result: 'Succeeded' })
    }

    public recordEnterFocusChat() {
        telemetry.amazonq_enterFocusChat.emit({ result: 'Succeeded' })
    }

    public recordExitFocusChat() {
        telemetry.amazonq_exitFocusChat.emit({ result: 'Succeeded' })
    }

    public async recordFeedback(message: ChatItemFeedbackMessage) {
        const logger = getLogger()
        try {
            await globals.telemetry.postFeedback({
                comment: JSON.stringify({
                    type: 'codewhisperer-chat-answer-feedback',
                    conversationId: this.getConversationId(message.tabID) ?? '',
                    messageId: message.messageId,
                    reason: message.selectedOption,
                    userComment: message.comment,
                }),
                sentiment: 'Negative',
            })
        } catch (err) {
            const errorMessage = (err as Error).message || 'Failed to submit feedback'
            logger.error(`CodeWhispererChat answer feedback failed: "Negative": ${errorMessage}`)

            this.recordFeedbackResult('Failed')

            return errorMessage
        }

        logger.info(`CodeWhispererChat answer feedback sent: "Negative"`)

        this.recordFeedbackResult('Succeeded')
    }

    public recordFeedbackResult(feedbackResult: Result) {
        telemetry.feedback_result.emit({ result: feedbackResult })
    }

    public recordInteractWithMessage(
        message:
            | InsertCodeAtCursorPosition
            | CopyCodeToClipboard
            | PromptMessage
            | ChatItemVotedMessage
            | SourceLinkClickMessage
            | ResponseBodyLinkClickMessage
            | FooterInfoLinkClick
    ) {
        const conversationId = this.getConversationId(message.tabID)
        let event: AmazonqInteractWithMessage | undefined
        switch (message.command) {
            case 'insert_code_at_cursor_position':
                message = message as InsertCodeAtCursorPosition
                event = {
                    cwsprChatConversationId: conversationId ?? '',
                    cwsprChatMessageId: message.messageId,
                    cwsprChatInteractionType: 'insertAtCursor',
                    cwsprChatAcceptedCharactersLength: message.code.length,
                    cwsprChatInteractionTarget: message.insertionTargetType,
                    cwsprChatHasReference: message.codeReference && message.codeReference.length > 0,
                }
                break
            case 'code_was_copied_to_clipboard':
                message = message as CopyCodeToClipboard
                event = {
                    cwsprChatConversationId: conversationId ?? '',
                    cwsprChatMessageId: message.messageId,
                    cwsprChatInteractionType: 'copySnippet',
                    cwsprChatAcceptedCharactersLength: message.code.length,
                    cwsprChatInteractionTarget: message.insertionTargetType,
                    cwsprChatHasReference: message.codeReference && message.codeReference.length > 0,
                }
                break
            case 'follow-up-was-clicked':
                message = message as PromptMessage
                event = {
                    cwsprChatConversationId: conversationId ?? '',
                    cwsprChatMessageId: message.messageId,
                    cwsprChatInteractionType: 'clickFollowUp',
                }
                break
            case 'chat-item-voted':
                message = message as ChatItemVotedMessage
                event = {
                    cwsprChatMessageId: message.messageId,
                    cwsprChatConversationId: conversationId ?? '',
                    cwsprChatInteractionType: message.vote,
                }
                break
            case 'source-link-click':
                message = message as SourceLinkClickMessage
                event = {
                    cwsprChatMessageId: message.messageId,
                    cwsprChatConversationId: conversationId ?? '',
                    cwsprChatInteractionType: 'clickLink',
                    cwsprChatInteractionTarget: message.link,
                }
                break
            case 'response-body-link-click':
                message = message as ResponseBodyLinkClickMessage
                event = {
                    cwsprChatMessageId: message.messageId,
                    cwsprChatConversationId: conversationId ?? '',
                    cwsprChatInteractionType: 'clickBodyLink',
                    cwsprChatInteractionTarget: message.link,
                }
                break
            case 'footer-info-link-click':
                message = message as FooterInfoLinkClick
                event = {
                    cwsprChatMessageId: 'footer',
                    cwsprChatConversationId: conversationId ?? '',
                    cwsprChatInteractionType: 'clickBodyLink',
                    cwsprChatInteractionTarget: message.link,
                }
                break
        }

        if (!event) {
            return
        }

        telemetry.amazonq_interactWithMessage.emit({ result: 'Succeeded', ...event })

        codeWhispererClient
            .sendTelemetryEvent(mapToClientTelemetryEvent('amazonq_interactWithMessage', event))
            .then()
            .catch(logSendTelemetryEventFailure)
    }

    public getTriggerInteractionFromTriggerEvent(triggerEvent: TriggerEvent | undefined): CwsprChatTriggerInteraction {
        switch (triggerEvent?.type) {
            case 'editor_context_command':
                return triggerEvent.command?.triggerType === 'keybinding' ? 'hotkeys' : 'contextMenu'
            case 'follow_up':
            case 'chat_message':
            default:
                return 'click'
        }
    }

    public recordStartConversation(triggerEvent: TriggerEvent, triggerPayload: TriggerPayload) {
        if (triggerEvent.tabID === undefined) {
            return
        }

        if (this.triggerEventsStorage.getTriggerEventsByTabID(triggerEvent.tabID).length > 1) {
            return
        }

        const telemetryUserIntent = this.getUserIntentForTelemetry(triggerPayload.userIntent)

        telemetry.amazonq_startConversation.emit({
            result: 'Succeeded',
            cwsprChatConversationId: this.getConversationId(triggerEvent.tabID) ?? '',
            cwsprChatTriggerInteraction: this.getTriggerInteractionFromTriggerEvent(triggerEvent),
            cwsprChatConversationType: 'Chat',
            cwsprChatUserIntent: telemetryUserIntent,
            cwsprChatHasCodeSnippet: triggerPayload.codeSelection && !triggerPayload.codeSelection.isEmpty,
            cwsprChatProgrammingLanguage: triggerPayload.fileLanguage,
        })
    }

    public recordAddMessage(triggerPayload: TriggerPayload, message: PromptAnswer) {
        const triggerEvent = this.triggerEventsStorage.getLastTriggerEventByTabID(message.tabID)

        const event: AmazonqAddMessage = {
            cwsprChatConversationId: this.getConversationId(message.tabID) ?? '',
            cwsprChatMessageId: message.messageID,
            cwsprChatTriggerInteraction: this.getTriggerInteractionFromTriggerEvent(triggerEvent),
            cwsprChatUserIntent: this.getUserIntentForTelemetry(triggerPayload.userIntent),
            cwsprChatHasCodeSnippet: triggerPayload.codeSelection && !triggerPayload.codeSelection.isEmpty,
            cwsprChatProgrammingLanguage: triggerPayload.fileLanguage,
            cwsprChatActiveEditorTotalCharacters: triggerPayload.fileText?.length,
            cwsprChatActiveEditorImportCount: triggerPayload.codeQuery?.fullyQualifiedNames?.used?.length,
            cwsprChatResponseCodeSnippetCount: 0, // TODO
            cwsprChatResponseCode: message.responseCode,
            cwsprChatSourceLinkCount: message.suggestionCount,
            cwsprChatReferencesCount: message.codeReferenceCount,
            cwsprChatFollowUpCount: message.followUpCount,
            cwsprChatTimeToFirstChunk: this.getResponseStreamTimeToFirstChunk(message.tabID),
            cwsprChatTimeBetweenChunks: '', // this.getResponseStreamTimeBetweenChunks(message.tabID), //TODO: allow '[', ']' and ',' chars
            cwsprChatFullResponseLatency: this.responseStreamTotalTime.get(message.tabID) ?? 0,
            cwsprChatRequestLength: triggerPayload.message?.length ?? 0,
            cwsprChatResponseLength: message.messageLength,
            cwsprChatConversationType: 'Chat',
        }

        telemetry.amazonq_addMessage.emit({ result: 'Succeeded', ...event })

        codeWhispererClient
            .sendTelemetryEvent(mapToClientTelemetryEvent('amazonq_addMessage', event))
            .then()
            .catch(logSendTelemetryEventFailure)
    }

    public recordMessageResponseError(triggerPayload: TriggerPayload, tabID: string, responseCode: number) {
        const triggerEvent = this.triggerEventsStorage.getLastTriggerEventByTabID(tabID)

        telemetry.amazonq_messageResponseError.emit({
            result: 'Succeeded',
            cwsprChatConversationId: this.getConversationId(tabID) ?? '',
            cwsprChatTriggerInteraction: this.getTriggerInteractionFromTriggerEvent(triggerEvent),
            cwsprChatUserIntent: this.getUserIntentForTelemetry(triggerPayload.userIntent),
            cwsprChatHasCodeSnippet: !triggerPayload.codeSelection?.isEmpty,
            cwsprChatProgrammingLanguage: triggerPayload.fileLanguage,
            cwsprChatActiveEditorTotalCharacters: triggerPayload.fileText?.length,
            cwsprChatActiveEditorImportCount: triggerPayload.codeQuery?.fullyQualifiedNames?.used?.length,
            cwsprChatResponseCode: responseCode,
            cwsprChatRequestLength: triggerPayload.message?.length ?? 0,
            cwsprChatConversationType: 'Chat',
        })
    }

    public recordEnterFocusConversation(tabID: string) {
        const conversationId = this.getConversationId(tabID)
        if (conversationId) {
            telemetry.amazonq_enterFocusConversation.emit({
                result: 'Succeeded',
                cwsprChatConversationId: conversationId,
            })
        }
    }

    public recordExitFocusConversation(tabID: string) {
        const conversationId = this.getConversationId(tabID)
        if (conversationId) {
            telemetry.amazonq_exitFocusConversation.emit({
                result: 'Succeeded',
                cwsprChatConversationId: conversationId,
            })
        }
    }

    public setResponseStreamStartTime(tabID: string) {
        this.responseStreamStartTime.set(tabID, performance.now())
        this.responseStreamTimeForChunks.set(tabID, [performance.now()])
    }

    public setResponseStreamTimeForChunks(tabID: string) {
        const chunkTimes = this.responseStreamTimeForChunks.get(tabID) ?? []
        this.responseStreamTimeForChunks.set(tabID, [...chunkTimes, performance.now()])
    }

    private getResponseStreamTimeToFirstChunk(tabID: string): number {
        const chunkTimes = this.responseStreamTimeForChunks.get(tabID) ?? [0, 0]
        if (chunkTimes.length === 1) {
            return Math.round(performance.now() - chunkTimes[0])
        }
        return Math.round(chunkTimes[1] - chunkTimes[0])
    }

    // private getResponseStreamTimeBetweenChunks(tabID: string): string {
    //     try {
    //         const chunkDeltaTimes: number[] = []
    //         const chunkTimes = this.responseStreamTimeForChunks.get(tabID) ?? [0]
    //         for (let idx = 0; idx < chunkTimes.length - 1; idx++) {
    //             chunkDeltaTimes.push(Math.round(chunkTimes[idx + 1] - chunkTimes[idx]))
    //         }

    //         const trimmed = JSON.stringify(chunkDeltaTimes).slice(0, 2_000)
    //         const fixed = trimmed.endsWith(',') ? trimmed.slice(0, -1) : trimmed
    //         return fixed.endsWith(']') ? fixed : `${fixed}]`
    //     } catch (e) {
    //         return '[-1]'
    //     }
    // }

    public setResponseStreamTotalTime(tabID: string) {
        const totalTime = performance.now() - (this.responseStreamStartTime.get(tabID) ?? 0)
        this.responseStreamTotalTime.set(tabID, Math.round(totalTime))
    }

    public getConversationId(tabID: string): string | undefined {
        return this.sessionStorage.getSession(tabID).sessionIdentifier
    }
}

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cwc/controller/chat/telemetry/TelemetryHelper.kt
// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package software.aws.toolkits.jetbrains.services.cwc.controller.chat.telemetry

import software.amazon.awssdk.services.codewhispererstreaming.model.UserIntent
import software.amazon.awssdk.services.toolkittelemetry.model.Sentiment
import software.aws.toolkits.core.utils.getLogger
import software.aws.toolkits.core.utils.info
import software.aws.toolkits.core.utils.warn
import software.aws.toolkits.jetbrains.services.amazonq.apps.AmazonQAppInitContext
import software.aws.toolkits.jetbrains.services.codewhisperer.credentials.CodeWhispererClientAdaptor
import software.aws.toolkits.jetbrains.services.cwc.clients.chat.model.ChatRequestData
import software.aws.toolkits.jetbrains.services.cwc.clients.chat.model.TriggerType
import software.aws.toolkits.jetbrains.services.cwc.controller.ChatController
import software.aws.toolkits.jetbrains.services.cwc.messages.ChatMessage
import software.aws.toolkits.jetbrains.services.cwc.messages.IncomingCwcMessage
import software.aws.toolkits.jetbrains.services.cwc.messages.LinkType
import software.aws.toolkits.jetbrains.services.cwc.storage.ChatSessionStorage
import software.aws.toolkits.jetbrains.services.telemetry.TelemetryService
import software.aws.toolkits.jetbrains.utils.notifyError
import software.aws.toolkits.resources.message
import software.aws.toolkits.telemetry.AmazonqTelemetry
import software.aws.toolkits.telemetry.CwsprChatCommandType
import software.aws.toolkits.telemetry.CwsprChatConversationType
import software.aws.toolkits.telemetry.CwsprChatInteractionType
import software.aws.toolkits.telemetry.CwsprChatTriggerInteraction
import software.aws.toolkits.telemetry.CwsprChatUserIntent
import software.aws.toolkits.telemetry.FeedbackTelemetry
import java.time.Duration
import java.time.Instant

class TelemetryHelper(private val context: AmazonQAppInitContext, private val sessionStorage: ChatSessionStorage) {

    private val responseStreamStartTime: MutableMap<String, Instant> = mutableMapOf()
    private val responseStreamTotalTime: MutableMap<String, Int> = mutableMapOf()
    private val responseStreamTimeForChunks: MutableMap<String, MutableList<Instant>> = mutableMapOf()

    fun getConversationId(tabId: String): String? = sessionStorage.getSession(tabId)?.session?.conversationId

    private fun getTelemetryUserIntent(userIntent: UserIntent): CwsprChatUserIntent = when (userIntent) {
        UserIntent.SUGGEST_ALTERNATE_IMPLEMENTATION -> CwsprChatUserIntent.SuggestAlternateImplementation
        UserIntent.APPLY_COMMON_BEST_PRACTICES -> CwsprChatUserIntent.ApplyCommonBestPractices
        UserIntent.IMPROVE_CODE -> CwsprChatUserIntent.ImproveCode
        UserIntent.SHOW_EXAMPLES -> CwsprChatUserIntent.ShowExample
        UserIntent.CITE_SOURCES -> CwsprChatUserIntent.CiteSources
        UserIntent.EXPLAIN_LINE_BY_LINE -> CwsprChatUserIntent.ExplainLineByLine
        UserIntent.EXPLAIN_CODE_SELECTION -> CwsprChatUserIntent.ExplainCodeSelection
        UserIntent.UNKNOWN_TO_SDK_VERSION -> CwsprChatUserIntent.Unknown
    }

    private fun getTelemetryTriggerType(triggerType: TriggerType): CwsprChatTriggerInteraction = when (triggerType) {
        TriggerType.Click -> CwsprChatTriggerInteraction.Click
        TriggerType.ContextMenu, TriggerType.Hotkeys -> CwsprChatTriggerInteraction.ContextMenu
    }

    // When chat panel is focused
    fun recordEnterFocusChat() {
        AmazonqTelemetry.enterFocusChat()
    }

    // When chat panel is unfocused
    fun recordExitFocusChat() {
        AmazonqTelemetry.exitFocusChat()
    }

    fun recordStartConversation(tabId: String, data: ChatRequestData) {
        val sessionHistory = sessionStorage.getSession(tabId)?.history ?: return
        if (sessionHistory.size > 1) return

        AmazonqTelemetry.startConversation(
            cwsprChatConversationId = getConversationId(tabId).orEmpty(),
            cwsprChatTriggerInteraction = getTelemetryTriggerType(data.triggerType),
            cwsprChatConversationType = CwsprChatConversationType.Chat,
            cwsprChatUserIntent = data.userIntent?.let { getTelemetryUserIntent(it) },
            cwsprChatHasCodeSnippet = data.activeFileContext.focusAreaContext?.codeSelection?.isNotEmpty() ?: false,
            cwsprChatProgrammingLanguage = data.activeFileContext.fileContext?.fileLanguage,
        )
    }

    // When Chat API responds to a user message (full response streamed)
    fun recordAddMessage(data: ChatRequestData, response: ChatMessage, responseLength: Int, statusCode: Int) {
        AmazonqTelemetry.addMessage(
            cwsprChatConversationId = getConversationId(response.tabId).orEmpty(),
            cwsprChatMessageId = response.messageId,
            cwsprChatTriggerInteraction = getTelemetryTriggerType(data.triggerType),
            cwsprChatUserIntent = data.userIntent?.let { getTelemetryUserIntent(it) },
            cwsprChatHasCodeSnippet = data.activeFileContext.focusAreaContext?.codeSelection?.isNotEmpty() ?: false,
            cwsprChatProgrammingLanguage = data.activeFileContext.fileContext?.fileLanguage,
            cwsprChatActiveEditorTotalCharacters = data.activeFileContext.focusAreaContext?.codeSelection?.length,
            cwsprChatActiveEditorImportCount = data.activeFileContext.focusAreaContext?.codeNames?.fullyQualifiedNames?.used?.size,
            cwsprChatResponseCodeSnippetCount = 0,
            cwsprChatResponseCode = statusCode,
            cwsprChatSourceLinkCount = response.relatedSuggestions?.size,
            cwsprChatReferencesCount = 0, // TODO
            cwsprChatFollowUpCount = response.followUps?.size,
            cwsprChatTimeToFirstChunk = getResponseStreamTimeToFirstChunk(response.tabId),
            cwsprChatTimeBetweenChunks = "", // getResponseStreamTimeBetweenChunks(response.tabId), //TODO: allow '[', ']' and ',' chars
            cwsprChatFullResponseLatency = responseStreamTotalTime[response.tabId] ?: 0,
            cwsprChatRequestLength = data.message.length,
            cwsprChatResponseLength = responseLength,
            cwsprChatConversationType = CwsprChatConversationType.Chat,
        )

        val metadata: Map<String, Any?> = mapOf(
            "cwsprChatConversationId" to getConversationId(response.tabId).orEmpty(),
            "cwsprChatMessageId" to response.messageId,
            "cwsprChatTriggerInteraction" to getTelemetryTriggerType(data.triggerType),
            "cwsprChatUserIntent" to data.userIntent?.let { getTelemetryUserIntent(it) },
            "cwsprChatHasCodeSnippet" to (data.activeFileContext.focusAreaContext?.codeSelection?.isNotEmpty() ?: false),
            "cwsprChatProgrammingLanguage" to data.activeFileContext.fileContext?.fileLanguage,
            "cwsprChatActiveEditorTotalCharacters" to data.activeFileContext.focusAreaContext?.codeSelection?.length,
            "cwsprChatActiveEditorImportCount" to data.activeFileContext.focusAreaContext?.codeNames?.fullyQualifiedNames?.used?.size,
            "cwsprChatResponseCodeSnippetCount" to 0,
            "cwsprChatResponseCode" to statusCode,
            "cwsprChatSourceLinkCount" to response.relatedSuggestions?.size,
            "cwsprChatReferencesCount" to 0, // TODO
            "cwsprChatFollowUpCount" to response.followUps?.size,
            "cwsprChatTimeToFirstChunk" to getResponseStreamTimeToFirstChunk(response.tabId),
            "cwsprChatTimeBetweenChunks" to "", // getResponseStreamTimeBetweenChunks(response.tabId), //TODO: allow '[', ']' and ',' chars
            "cwsprChatFullResponseLatency" to (responseStreamTotalTime[response.tabId] ?: 0),
            "cwsprChatRequestLength" to data.message.length,
            "cwsprChatResponseLength" to responseLength,
            "cwsprChatConversationType" to CwsprChatConversationType.Chat
        )
        sendMetricData("amazonq_addMessage", metadata)
    }

    fun recordMessageResponseError(data: ChatRequestData, tabId: String, responseCode: Int) {
        AmazonqTelemetry.messageResponseError(
            cwsprChatConversationId = getConversationId(tabId).orEmpty(),
            cwsprChatTriggerInteraction = getTelemetryTriggerType(data.triggerType),
            cwsprChatUserIntent = data.userIntent?.let { getTelemetryUserIntent(it) },
            cwsprChatHasCodeSnippet = data.activeFileContext.focusAreaContext?.codeSelection?.isNotEmpty() ?: false,
            cwsprChatProgrammingLanguage = data.activeFileContext.fileContext?.fileLanguage,
            cwsprChatActiveEditorTotalCharacters = data.activeFileContext.focusAreaContext?.codeSelection?.length,
            cwsprChatActiveEditorImportCount = data.activeFileContext.focusAreaContext?.codeNames?.fullyQualifiedNames?.used?.size,
            cwsprChatResponseCode = responseCode,
            cwsprChatRequestLength = data.message.length,
            cwsprChatConversationType = CwsprChatConversationType.Chat,
        )
    }

    // When user interacts with a message (e.g. copy code, insert code, vote)
    suspend fun recordInteractWithMessage(message: IncomingCwcMessage) {
        val metadata: Map<String, Any?> = when (message) {
            is IncomingCwcMessage.ChatItemVoted -> {
                AmazonqTelemetry.interactWithMessage(
                    cwsprChatConversationId = getConversationId(message.tabId).orEmpty(),
                    cwsprChatMessageId = message.messageId,
                    cwsprChatInteractionType = when (message.vote) {
                        "upvote" -> CwsprChatInteractionType.Upvote
                        "downvote" -> CwsprChatInteractionType.Downvote
                        else -> CwsprChatInteractionType.Unknown
                    },
                )
                mapOf(
                    "cwsprChatConversationId" to getConversationId(message.tabId).orEmpty(),
                    "cwsprChatMessageId" to message.messageId,
                    "cwsprChatInteractionType" to when (message.vote) {
                        "upvote" -> CwsprChatInteractionType.Upvote
                        "downvote" -> CwsprChatInteractionType.Downvote
                        else -> CwsprChatInteractionType.Unknown
                    }
                )
            }

            is IncomingCwcMessage.FollowupClicked -> {
                AmazonqTelemetry.interactWithMessage(
                    cwsprChatConversationId = getConversationId(message.tabId).orEmpty(),
                    cwsprChatMessageId = message.messageId.orEmpty(),
                    cwsprChatInteractionType = CwsprChatInteractionType.ClickFollowUp,
                )
                mapOf(
                    "cwsprChatConversationId" to getConversationId(message.tabId).orEmpty(),
                    "cwsprChatMessageId" to message.messageId.orEmpty(),
                    "cwsprChatInteractionType" to CwsprChatInteractionType.ClickFollowUp,
                )
            }

            is IncomingCwcMessage.CopyCodeToClipboard -> {
                AmazonqTelemetry.interactWithMessage(
                    cwsprChatConversationId = getConversationId(message.tabId).orEmpty(),
                    cwsprChatMessageId = message.messageId,
                    cwsprChatInteractionType = CwsprChatInteractionType.CopySnippet,
                    cwsprChatAcceptedCharactersLength = message.code.length,
                    cwsprChatInteractionTarget = message.insertionTargetType,
                    cwsprChatHasReference = null,
                )
                mapOf(
                    "cwsprChatConversationId" to getConversationId(message.tabId).orEmpty(),
                    "cwsprChatMessageId" to message.messageId,
                    "cwsprChatInteractionType" to CwsprChatInteractionType.CopySnippet,
                    "cwsprChatAcceptedCharactersLength" to message.code.length,
                    "cwsprChatInteractionTarget" to message.insertionTargetType,
                    "cwsprChatHasReference" to null,
                )
            }

            is IncomingCwcMessage.InsertCodeAtCursorPosition -> {
                AmazonqTelemetry.interactWithMessage(
                    cwsprChatConversationId = getConversationId(message.tabId).orEmpty(),
                    cwsprChatMessageId = message.messageId,
                    cwsprChatInteractionType = CwsprChatInteractionType.InsertAtCursor,
                    cwsprChatAcceptedCharactersLength = message.code.length,
                    cwsprChatInteractionTarget = message.insertionTargetType,
                    cwsprChatHasReference = null,
                )
                mapOf(
                    "cwsprChatConversationId" to getConversationId(message.tabId).orEmpty(),
                    "cwsprChatMessageId" to message.messageId,
                    "cwsprChatInteractionType" to CwsprChatInteractionType.InsertAtCursor,
                    "cwsprChatAcceptedCharactersLength" to message.code.length,
                    "cwsprChatInteractionTarget" to message.insertionTargetType,
                    "cwsprChatHasReference" to null,
                )
            }

            is IncomingCwcMessage.ClickedLink -> {
                // Null when internal Amazon link is clicked
                if (message.messageId == null) return

                val linkInteractionType = when (message.type) {
                    LinkType.SourceLink -> CwsprChatInteractionType.ClickLink
                    LinkType.BodyLink -> CwsprChatInteractionType.ClickBodyLink
                    else -> CwsprChatInteractionType.Unknown
                }
                AmazonqTelemetry.interactWithMessage(
                    cwsprChatConversationId = getConversationId(message.tabId).orEmpty(),
                    cwsprChatMessageId = message.messageId,
                    cwsprChatInteractionType = linkInteractionType,
                    cwsprChatInteractionTarget = message.link,
                    cwsprChatHasReference = null,
                )
                mapOf(
                    "cwsprChatConversationId" to getConversationId(message.tabId).orEmpty(),
                    "cwsprChatMessageId" to message.messageId,
                    "cwsprChatInteractionType" to linkInteractionType,
                    "cwsprChatInteractionTarget" to message.link,
                    "cwsprChatHasReference" to null,
                )
            }

            is IncomingCwcMessage.ChatItemFeedback -> {
                recordFeedback(message)
                emptyMap()
            }

            else -> emptyMap()
        }

        sendMetricData("amazonq_interactWithMessage", metadata)
    }

    private suspend fun recordFeedback(message: IncomingCwcMessage.ChatItemFeedback) {
        val comment = FeedbackComment(
            conversationId = getConversationId(message.tabId).orEmpty(),
            messageId = message.messageId,
            reason = message.selectedOption,
            userComment = message.comment.orEmpty(),
        )

        try {
            TelemetryService.getInstance().sendFeedback(
                sentiment = Sentiment.NEGATIVE,
                comment = ChatController.objectMapper.writeValueAsString(comment),
            )
            logger.info { "CodeWhispererChat answer feedback sent: \"Negative\"" }
            recordFeedbackResult(true)
        } catch (e: Throwable) {
            e.notifyError(message("feedback.submit_failed", e))
            logger.warn(e) { "Failed to submit feedback" }
            recordFeedbackResult(false)
            return
        }
    }

    private fun recordFeedbackResult(success: Boolean) {
        FeedbackTelemetry.result(project = null, success = success)
    }

    // When a conversation(tab) is focused
    fun recordEnterFocusConversation(tabId: String) {
        getConversationId(tabId)?.let {
            AmazonqTelemetry.enterFocusConversation(
                cwsprChatConversationId = it,
            )
        }
    }

    // When a conversation(tab) is unfocused
    fun recordExitFocusConversation(tabId: String) {
        getConversationId(tabId)?.let {
            AmazonqTelemetry.exitFocusConversation(
                cwsprChatConversationId = it,
            )
        }
    }

    fun setResponseStreamStartTime(tabId: String) {
        responseStreamStartTime[tabId] = Instant.now()
        responseStreamTimeForChunks[tabId] = mutableListOf(Instant.now())
    }

    fun setResponseStreamTimeForChunks(tabId: String) {
        val chunkTimes = responseStreamTimeForChunks.getOrPut(tabId) { mutableListOf() }
        chunkTimes += Instant.now()
    }

    fun setResponseStreamTotalTime(tabId: String) {
        val totalTime = Duration.between(responseStreamStartTime[tabId], Instant.now()).toMillis().toInt()
        responseStreamTotalTime[tabId] = totalTime
    }

    private fun getResponseStreamTimeToFirstChunk(tabId: String): Int {
        val chunkTimes = responseStreamTimeForChunks[tabId] ?: return 0
        if (chunkTimes.size == 1) return Duration.between(chunkTimes[0], Instant.now()).toMillis().toInt()
        return Duration.between(chunkTimes[0], chunkTimes[1]).toMillis().toInt()
    }

//    private fun getResponseStreamTimeBetweenChunks(tabId: String): String = try {
//        val chunkDeltaTimes = mutableListOf<Int>()
//        val chunkTimes = responseStreamTimeForChunks[tabId] ?: listOf(Instant.now())
//        for (idx in 0 until (chunkTimes.size - 1)) {
//            chunkDeltaTimes += Duration.between(chunkTimes[idx], chunkTimes[idx + 1]).toMillis().toInt()
//        }
//
//        val joined = "[" + chunkDeltaTimes.joinToString(",").take(2000)
//        val fixed = if (joined.last() == ',') "${joined}0" else joined
//        if (fixed.last() == ']') fixed else "$fixed]"
//    } catch (e: Exception) {
//        "[-1]"
//    }

    private fun sendMetricData(eventName: String, metadata: Map<String, Any?>) {
        try {
            CodeWhispererClientAdaptor.getInstance(context.project).sendMetricDataTelemetry(eventName, metadata)
        } catch (e: Throwable) {
            logger.warn(e) { "Failed to send metric data" }
        }
    }

    companion object {
        private val logger = getLogger<TelemetryHelper>()

        fun recordOpenChat() {
            AmazonqTelemetry.openChat()
        }

        fun recordCloseChat() {
            AmazonqTelemetry.closeChat()
        }

        fun recordTelemetryChatRunCommand(type: CwsprChatCommandType, name: String? = null) {
            AmazonqTelemetry.runCommand(cwsprChatCommandType = type, cwsprChatCommandName = name)
        }
    }
}

data class FeedbackComment(
    val conversationId: String,
    val messageId: String,
    val reason: String,
    val userComment: String,
    val type: String = "codewhisperer-chat-answer-feedback",
)
